<h1>Project: United Kingdom Road Accident Data Analysis</h1>
<h2>Inclusive Years: 2019 - 2022</h2>
<h3> Analyst : Sophia Joyce Lozada </h3>

In [1]:
pip install numpy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC11\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [2]:

pip install pandas

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC11\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [3]:
pip install matplotlib

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC11\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [4]:
pip install seaborn

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC11\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [5]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt


In [6]:
pip install scipy


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC11\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [7]:
from scipy.stats import f_oneway

In [8]:
import warnings
warnings.filterwarnings('ignore')

<h1> Data Analytics</h1>
<h2>United Kingdom Road Accident Data Analysis</h2>

In [9]:
accident = pd.read_csv ('dataset\\accident_data.csv')

In [10]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [11]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [12]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<h1>Check for Missing Values</h1>

In [13]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [14]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unaccounted')
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')

In [15]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1> Converting the Data Types </h1>

In [16]:
accident['Vehicle_Type'] =accident['Vehicle_Type'].astype('category')
accident['Latitude'] =accident['Latitude'].astype('category')
accident['Longitude'] =accident['Longitude'].astype('category')
accident['Light_Conditions'] =accident['Light_Conditions'].astype('category')
accident['District Area'] =accident['District Area'].astype('category')
accident['Road_Type'] =accident['Road_Type'].astype('category')
accident['Index'] =accident['Index'].astype('category')
accident['Accident_Severity'] =accident['Accident_Severity'].astype('category')
accident['Road_Surface_Conditions'] =accident['Road_Surface_Conditions'].astype('category')
accident['Urban_or_Rural_Area'] =accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] =accident['Weather_Conditions'].astype('category')


In [17]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  object  
 3   Latitude                 660679 non-null  category
 4   Light_Conditions         660679 non-null  category
 5   District Area            660679 non-null  category
 6   Longitude                660679 non-null  category
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                660679 non-null  category
 11  Urban_or_Rural_Area      660679 non-null  category
 12  Weather_Conditions       660679 non-null  category
 13  Vehicle_Type             660679 non-null  ca

<h1>Converting Object to DateTime Data Type</h1>

In [18]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'] , dayfirst = True , errors = 'coerce')


In [19]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Extracting date information using pandas date time</h1>

In [20]:
accident['Year'] =accident['Accident Date'].dt.year
accident['Month'] =accident['Accident Date'].dt.month
accident['Day'] =accident['Accident Date'].dt.day
accident['Day of the Week '] =accident['Accident Date'].dt.dayofweek

<h1>DATA ANALYTICS</h1>
<h3>analyzing each field from the data set</h3>

<h1>UniVariate</h1>
<h3>Univariate Analysis from year 2019 - 2022</h3>
<h3>Question: How many incidents happened from year 2019 -2022  under light conditions?</h3>

In [56]:
df2019= accident[accident['Year'] == 2019]
df2020= accident[accident['Year'] == 2020]
df2021= accident[accident['Year'] == 2021]
df2022= accident[accident['Year'] == 2022]


In [22]:
df2019['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

In [23]:
df2019['Light_Conditions'].value_counts().sum()

np.int64(182115)

<h1>YEAR 2019</h1>
<u>
    <li>  73% of the accidents happened in Daylight </li>
    <li>  19% of the accidents happened in Darkness - lights lit</li>
    <li>  6% of the accidents happened in Darkness - no lighting</li>
    <li> 0.97% of the accidents happened in Darkness - lighting unknown</li>
    <li> 0.38% of the accidents happened in Darkness - lights unlit</li>
</u>
    

In [24]:
df2020['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       124552
Darkness - lights lit           33473
Darkness - no lighting          10067
Darkness - lighting unknown      1792
Darkness - lights unlit           707
Name: count, dtype: int64

In [25]:
df2020['Light_Conditions'].value_counts().sum()

np.int64(170591)

<h1>YEAR 2020</h1>
<u>
    <li>  73.01% of the accidents happened in Daylight </li>
    <li>  19.62% of the accidents happened in Darkness - lights lit</li>
    <li> 5.90% of the accidents happened in Darkness - no lighting</li>
    <li> 1.05% of the accidents happened in Darkness - lighting unknown</li>
    <li> 0.41% of the accidents happened in Darkness - lights unlit</li>
</u>

In [26]:
df2021['Light_Conditions'].value_counts().sum()

np.int64(163554)

In [27]:
df2021['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       119518
Darkness - lights lit           32549
Darkness - no lighting           9226
Darkness - lighting unknown      1607
Darkness - lights unlit           654
Name: count, dtype: int64

In [28]:
df2021['Light_Conditions'].value_counts().sum()

np.int64(163554)

<h1>YEAR 2021</h1>
<u>
    <li> 73.07% of the accidents happened in Daylight </li>
    <li>  19.90% of the accidents happened in Darkness - lights lit</li>
    <li> 5.64% of the accidents happened in Darkness - no lighting</li>
    <li> 0.98% of the accidents happened in Darkness - lighting unknown</li>
    <li> 0.39% of the accidents happened in Darkness - lights unlit</li>
</u>

In [29]:
df2022['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       107768
Darkness - lights lit           27544
Darkness - no lighting           7302
Darkness - lighting unknown      1317
Darkness - lights unlit           488
Name: count, dtype: int64

In [30]:
df2022['Light_Conditions'].value_counts().sum()

np.int64(144419)

<h1>YEAR 2022</h1>
<u>
    <li> 74.62% of the accidents happened in Daylight </li>
    <li>  19.07% of the accidents happened in Darkness - lights lit</li>
    <li>  5.05% of the accidents happened in Darkness - no lighting</li>
    <li> 0.91% of the accidents happened in Darkness - lighting unknown</li>
    <li> 0.33% of the accidents happened in Darkness - lights unlit</li>
</u>

<h1>Insight No.1 - Daylight consistently has the highest number of incidents across all years</h1>
<h1>Insight No.2 - Darkness - lights lit is the second most common condition </h1>
<h1>Insight No.3 - Darkness - lights unlit has the lowest counts and also declines over time</h1>

In [50]:
accident['Accident Date'].dt.month.value_counts()

Accident Date
11    60424
10    59580
7     57445
6     56481
9     56455
5     56352
3     54086
8     53913
1     52872
12    51836
4     51744
2     49491
Name: count, dtype: int64

<h1>Based on the results , NOVEMBER has the highest number of recorded accidents.</h1>

In [77]:
m2019= accident[accident['Month'] == 2019]
m2020= accident[accident['Month'] == 2020]
m2021= accident[accident['Month'] == 2021]
m2022= accident[accident['Month'] == 2022] 

<h1>Insight No. 4 - Top 5 accident dates with the highest number of accidents</h1>

In [31]:
accident['Accident Date'].value_counts().head(5)

Accident Date
2019-11-30    704
2019-01-31    697
2019-07-13    692
2021-11-13    692
2019-08-14    688
Name: count, dtype: int64

<h1>Insight No.5 - Number of accidents per year </h1>
<h3> in year 2019  , the recorded accident is  182,115 </h3>
<h3> in year 2020  , the recorded accident is  170,591 </h3>
<h3> in year 2021  , the recored accident  is  163,554 </h3>
<h3> in year 2022  , the recored accident  is  144,419</h3>

In [32]:
accident['Accident Date'].dt.year.value_counts()

Accident Date
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<h1>Insight No.6 - How many accidents are recorded? </h1>
<h3> "the total accidents recorded is 660,679 "</h3>

In [33]:
accident.shape[0]

660679

<h1>Insight No.7 -Most common accident severity </h1>
<h3>"slight"</h3>

In [34]:
accident['Accident_Severity'].mode()[0]

'Slight'

<h1>Insight No.8 - Number of accidents in urban vs rural areas</h1>
<h3> "421,678 accidents occurred in urban areas while 238,990 accidents occurred in rural areas"</h3>

In [35]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h1>Insight No.9 - Most common weather conditions during accidents</h1>
<h3>Based on the data , the most common weather are fine no high winds and raining no high winds </h3>

In [36]:
accident['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unaccounted               14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h1>Insight No.10 - Average number of casualties per accident</h1>

In [67]:
accident['Number_of_Casualties'].mean()

np.float64(1.357040257068864)

<h1>Insight No.11 - Highest number of casualties recorded in a single accident</h1>

In [68]:
accident['Number_of_Casualties'].max()

np.int64(68)

<h1>Insight No.12 - Total number of vehicles involved in all accidents </h1>

In [69]:
accident['Number_of_Vehicles'].sum()

np.int64(1209872)

<h1>Insight No.13 - Most common vehicle type in accidents </h1>
<h3>"car"</h3>

In [ ]:
accident['Vehicle_Type'].mode()[0]

<h1>Insight No.14 - Most dangerous weather condition for accidents </h1>

In [63]:
accident.groupby("Weather_Conditions")["Accident_Severity"].value_counts()

Weather_Conditions     Accident_Severity
Fine + high winds      Slight                 7134
                       Serious                1245
                       Fatal                   175
Fine no high winds     Slight               441739
                       Serious               72046
                       Fatal                  7100
Fog or mist            Slight                 2963
                       Serious                 483
                       Fatal                    82
Other                  Slight                15184
                       Serious                1801
                       Fatal                   165
Raining + high winds   Slight                 8209
                       Serious                1261
                       Fatal                   145
Raining no high winds  Slight                69380
                       Serious                9468
                       Fatal                   848
Snowing + high winds   Slight            

<h1>Insight No.15 - Find the district with the highest number of casualties recorded </h1>
<h3>" Birmingham "</h3>

In [ ]:
accident.groupby("District Area")["Number_of_Casualties"].sum().idxmax()

<h1>Insight No.16 - Average accident severity by road type</h1>

In [62]:
accident.groupby("Road_Type")["Accident_Severity"].value_counts()

Road_Type           Accident_Severity
Dual carriageway    Slight                85863
                    Serious               11746
                    Fatal                  1815
One way street      Slight                11809
                    Serious                1655
                    Fatal                    95
Roundabout          Slight                40185
                    Serious                3665
                    Fatal                   142
Single carriageway  Slight               415557
                    Serious               70059
                    Fatal                  6527
Slip road           Slight                 6381
                    Serious                 611
                    Fatal                    49
unaccounted         Slight                 4006
                    Serious                 481
                    Fatal                    33
Name: count, dtype: int64

<h1>Insight No.17 - District with the Fewest Accidents</h1>
<h3>" Clackmannanshire "</h3>

In [ ]:
accident["District Area"].value_counts().idxmin()

<h1>Insight No.18 - Average Number of Casualties in Urban vs. Rural Areas</h1>

In [70]:
urban_casualties = accident[accident["Urban_or_Rural_Area"] == "Urban"]["Number_of_Casualties"].mean()
rural_casualties = accident[accident["Urban_or_Rural_Area"] == "Rural"]["Number_of_Casualties"].mean()

In [71]:
urban_casualties

np.float64(1.2878072842310957)

In [72]:
rural_casualties

np.float64(1.4792041508012888)

<h1>Insight No.19 - Highest Number of Vehicles in a Single Accident</h1>

In [73]:
max_vehicles = accident["Number_of_Vehicles"].max()
max_vehicles

np.int64(32)

<h1>Insight No.20 -  Accidents with More Than 5 Casualties</h1>

In [74]:
high_casualty_accidents = (accident["Number_of_Casualties"] > 5).sum()
high_casualty_accidents

np.int64(2612)

<h1>Insight No.21 -</h1>